src/nvim/getchar.c

```diff
@@ -1644,6 +1644,342 @@ static int handle_int(const int advance)
 
   return c;
 }
+
+// Returns 0 to continue, -1 to carry on in the loop, otherwise, returns the
+// character that should be used.
+// Keeps track of the mapdepth in the variable pointed to by "mapdepthp".
+static int look_in_typebuf(int *mapdepthp, int *keylenp,
+    int *mp_match_lenp,
+    const int timedout, const int advance, const int local_State)
```

Comment:
This function is coupled with the current state of the world and is used only once. It's not pure and it's not reusable. Finding a good name for this function is impossible, hence the `look_in_typebuf` name.

The existing code is horribly complicated but keeping it inline is better than extracting it.